internal/controller/clusterorder_webhook.go

```diff
@@ -6,40 +6,102 @@ import (
 
 	cloudkitv1alpha1 "github.com/innabox/cloudkit-operator/api/v1alpha1"
 )
 
-func triggerWebHook(ctx context.Context, url string, instance *cloudkitv1alpha1.ClusterOrder) error {
+type InflightRequest struct {
+	createTime time.Time
+}
+
+var (
+	inflightRequestsLock *sync.RWMutex              = &sync.RWMutex{}
+	inflightRequests     map[string]InflightRequest = make(map[string]InflightRequest)
+)
+
+func checkForExistingRequest(url string, minimumRequestInterval time.Duration) time.Duration {
+	var delta time.Duration
+
+	inflightRequestsLock.RLock()
+	request, ok := inflightRequests[url]
+	if ok {
+		delta = time.Since(request.createTime)
+		if delta >= minimumRequestInterval {
+			delta = 0
+		}
+	}
+	inflightRequestsLock.RUnlock()
+	purgeExpiredRequests(minimumRequestInterval)
+	return delta
+}
+
+func addInflightRequest(url string, minimumRequestInterval time.Duration) {
+	inflightRequestsLock.Lock()
+	inflightRequests[url] = InflightRequest{
+		createTime: time.Now(),
+	}
+	inflightRequestsLock.Unlock()
+	purgeExpiredRequests(minimumRequestInterval)
+}
+
+func purgeExpiredRequests(minimumRequestInterval time.Duration) {
+	var expiredRequests = make([]string, 3)
+
+	inflightRequestsLock.RLock()
+	for url, request := range inflightRequests {
+		if delta := time.Since(request.createTime); delta > minimumRequestInterval {
+			expiredRequests = append(expiredRequests, url)
+		}
+	}
+	inflightRequestsLock.RUnlock()
+
+	if len(expiredRequests) > 0 {
+		inflightRequestsLock.Lock()
+		for url := range inflightRequests {
+			delete(inflightRequests, url)
+		}
+		inflightRequestsLock.Unlock()
+	}
+}
+
+func triggerWebHook(ctx context.Context, url string, instance *cloudkitv1alpha1.ClusterOrder, minimumRequestInterval time.Duration) (time.Duration, error) {
+
 	log := ctrllog.FromContext(ctx)
+
+	if delta := checkForExistingRequest(url, minimumRequestInterval); delta != 0 {
+		return delta, nil
+	}
+
 	log.Info("Triggering webhook " + url)
 
 	jsonData, err := json.Marshal(instance)
 	if err != nil {
-		return fmt.Errorf("failed to marshal JSON: %w", err)
+		return 0, fmt.Errorf("failed to marshal JSON: %w", err)
 	}
 
 	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
 	if err != nil {
-		return fmt.Errorf("failed to create request: %w", err)
+		return 0, fmt.Errorf("failed to create request: %w", err)
 	}
 
 	req.Header.Set("Content-Type", "application/json")
 
 	client := &http.Client{Timeout: 10 * time.Second}
 	resp, err := client.Do(req)
 	if err != nil {
-		return fmt.Errorf("failed to send request: %w", err)
+		return 0, fmt.Errorf("failed to send request: %w", err)
 	}
 	defer resp.Body.Close() //nolint:errcheck
 
 	// Check response status
 	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
-		return fmt.Errorf("received non-success status code: %d", resp.StatusCode)
+		return 0, fmt.Errorf("received non-success status code: %d", resp.StatusCode)
 	}
 
-	return nil
+	addInflightRequest(url, minimumRequestInterval)
```

Comment:
This never gets called -- we exit above for successful calls to the webhook, so you're never adding URLs to inflightRequests and no rate limiting is being applied.

Reply:
I added some additional logging; it looks like urls are being purged from `inflightRequests` immediately after they are added.